WordPress/Classes/Utility/Media/FaviconService.swift

```diff
@@ -0,0 +1,107 @@
+import UIKit
+
+// Fetches URLs for favicons for sites.
+actor FaviconService {
+    static let shared = FaviconService()
+
+    private nonisolated let cache = FaviconCache()
+
+    private let session = URLSession(configuration: {
+        let configuration = URLSessionConfiguration.default
+        configuration.urlCache = nil
+        return configuration
+    }())
+
+    private var tasks: [URL: FaviconTask] = [:]
+
+    nonisolated func cachedFavicon(forURL siteURL: URL) -> URL? {
+        cache.cachedFavicon(forURL: siteURL)
+    }
+
+    /// Returns a favicon URL for the given site.
+    func favicon(forURL siteURL: URL) async throws -> URL {
+        if let faviconURL = cache.cachedFavicon(forURL: siteURL) {
+            return faviconURL
+        }
+        let faviconURL = try await _favicon(forURL: siteURL)
+        cache.storeCachedFaviconURL(faviconURL, forURL: siteURL)
+        return faviconURL
+    }
+
+    private func _favicon(forURL siteURL: URL) async throws -> URL {
+        let task = tasks[siteURL] ?? FaviconTask { [session] in
+            let (data, response) = try await session.data(from: siteURL)
+            try validate(response: response)
+            return await makeFavicon(from: data, siteURL: siteURL)
+        }
+        let subscriptionID = UUID()
+        task.subscriptions.insert(subscriptionID)
+        tasks[siteURL] = task
+        return try await withTaskCancellationHandler {
+            try await task.task.value
+        } onCancel: {
+            Task {
+                await self.unsubscribe(subscriptionID, key: siteURL)
+            }
+        }
+    }
```

Comment:
This function seems very carefully coded. What do you think about adding a few simple unit tests to prevent future us from breaking it?
1. Call `favicon(forURL:)` from multiple threads concurrently and make sure there is only one HTTP request sent out.
2. `Task { favicon(forURL:) }.cancel()` cancels the HTTP request.

Reply:
I'm not sure we'll continue using this service, so I don't think it's worth investing time into it right now. The task coalescing implementation is copied from `ImageDownloader` that has (some) tests.